{
 "metadata": {
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  },
  "orig_nbformat": 4,
  "kernelspec": {
   "name": "python3",
   "display_name": "Python 3.8.5 64-bit ('basicsynbio': pyenv)"
  },
  "interpreter": {
   "hash": "57f97546b8e059ecb785e5a83a832483957e404968430921b0ed5f9de2bd71dc"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2,
 "cells": [
  {
   "source": [
    "# Integration of basicsynbio and DNA chisel\n",
    "\n",
    "This notebook explores the integration of DNA chisel into basicsynbio for linker design purposes\n",
    "\n",
    "## Aims and objectives for cell/s below\n",
    "\n",
    "- [x] Try out DNA Chisel with easy to implement constraints.\n",
    "- [x] Make a Bowtie 2 file for sequences that will be present in the basicsynbio PartLinkerCollections.\n",
    "- [x] Run DNA Chisel to generate the backbone linker for the addgene collection."
   ],
   "cell_type": "markdown",
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "from dnachisel import (\n",
    "    AvoidChanges,\n",
    "    AvoidMatches,\n",
    "    AvoidPattern,\n",
    "    DnaOptimizationProblem,\n",
    "    EnforceGCContent,\n",
    "    EnforceMeltingTemperature,\n",
    "    EnzymeSitePattern,\n",
    "    random_dna_sequence,\n",
    ")\n",
    "import basicsynbio as bsb\n",
    "from Bio import (\n",
    "    Entrez,\n",
    "    SeqIO\n",
    ")\n",
    "from pathlib import Path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "constraint:   0%|          | 0/21 [00:00<?, ?it/s, now=AvoidPattern[0-55](patter...]\n",
      "\n",
      "location:   0%|          | 0/1 [00:00<?, ?it/s, now=None]\u001b[A\u001b[A\n",
      "\n",
      "location:   0%|          | 0/1 [00:00<?, ?it/s, now=36-39(+)]\u001b[A\u001b[A\n",
      "\n",
      "                                                                                    ===> SUCCESS - all constraints evaluations pass\n",
      "✔PASS ┍ EnforceMeltingTemperature[13-34]\n",
      "      │ Tm = 52.9\n",
      "✔PASS ┍ AvoidPattern[0-55](pattern:EcoRI(GAATTC))\n",
      "      │ Passed. Pattern not found !\n",
      "✔PASS ┍ AvoidPattern[0-55](pattern:SpeI(ACTAGT))\n",
      "      │ Passed. Pattern not found !\n",
      "✔PASS ┍ AvoidPattern[0-55](pattern:XbaI(TCTAGA))\n",
      "      │ Passed. Pattern not found !\n",
      "✔PASS ┍ AvoidPattern[0-55](pattern:PstI(CTGCAG))\n",
      "      │ Passed. Pattern not found !\n",
      "✔PASS ┍ AvoidPattern[0-55](pattern:BsaI(GGTCTC))\n",
      "      │ Passed. Pattern not found !\n",
      "✔PASS ┍ AvoidPattern[0-55](pattern:BsmBI(CGTCTC))\n",
      "      │ Passed. Pattern not found !\n",
      "✔PASS ┍ AvoidPattern[0-55](pattern:TTGACA)\n",
      "      │ Passed. Pattern not found !\n",
      "✔PASS ┍ AvoidPattern[0-55](pattern:TATAAT)\n",
      "      │ Passed. Pattern not found !\n",
      "✔PASS ┍ AvoidPattern[0-55](pattern:TTGNNNNNNNNNNNNNNNNNNNNTATNNT)\n",
      "      │ Passed. Pattern not found !\n",
      "✔PASS ┍ AvoidPattern[0-55](pattern:TGGCACGNNNNTTGC)\n",
      "      │ Passed. Pattern not found !\n",
      "✔PASS ┍ AvoidPattern[0-55](pattern:GAACTNNNNNNNNNNNNNNNNGTCNNA)\n",
      "      │ Passed. Pattern not found !\n",
      "✔PASS ┍ AvoidPattern[0-55](pattern:AAAGA)\n",
      "      │ Passed. Pattern not found !\n",
      "✔PASS ┍ AvoidPattern[0-55](pattern:AGGAGG)\n",
      "      │ Passed. Pattern not found !\n",
      "✔PASS ┍ AvoidPattern[0-55](pattern:ATG)\n",
      "      │ Passed. Pattern not found !\n",
      "✔PASS ┍ AvoidPattern[0-55](pattern:TTATNCACA)\n",
      "      │ Passed. Pattern not found !\n",
      "✔PASS ┍ AvoidPattern[0-55](pattern:TGTGANNNNNNTCACANT)\n",
      "      │ Passed. Pattern not found !\n",
      "✔PASS ┍ AvoidPattern[0-55](pattern:NGCTNAGCN)\n",
      "      │ Passed. Pattern not found !\n",
      "✔PASS ┍ AvoidPattern[0-55](pattern:GGGNNNNNCCC)\n",
      "      │ Passed. Pattern not found !\n",
      "✔PASS ┍ AvoidPattern[0-55](pattern:(G{3,}[ATGC]{1,7}){3,}G{3,})\n",
      "      │ Passed. Pattern not found !\n",
      "✔PASS ┍ AvoidPattern[0-55](pattern:GGGG)\n",
      "      │ Passed. Pattern not found !\n",
      "✔PASS ┍ AvoidChanges[0-6(+)]\n",
      "      │ Enforced by nucleotides restrictions\n",
      "✔PASS ┍ AvoidChanges[51-55(+)]\n",
      "      │ Enforced by nucleotides restrictions\n",
      "\n",
      "\n",
      "GGCTCGCTCCACCTGCGTCTATCGTACAGTGCTAAAATGGCAGCAGATAGTGTCC\n",
      "GGCTCGCTCCACCTGCGTCTATCGTACAGTGCTAAAATCGCAGCAGATAGTGTCC\n"
     ]
    }
   ],
   "source": [
    "linker_base_sequence = \"GGCTCG\" + random_dna_sequence(45, seed=123) + \"GTCC\"\n",
    "constraints = [\n",
    "    EnforceMeltingTemperature(mini=50, maxi=65, location=(13, 34)),\n",
    "    AvoidPattern(EnzymeSitePattern(\"EcoRI\")),\n",
    "    AvoidPattern(EnzymeSitePattern(\"SpeI\")),\n",
    "    AvoidPattern(EnzymeSitePattern(\"XbaI\")),\n",
    "    AvoidPattern(EnzymeSitePattern(\"PstI\")),\n",
    "    AvoidPattern(EnzymeSitePattern(\"BsaI\")),\n",
    "    AvoidPattern(EnzymeSitePattern(\"BsmBI\")),\n",
    "    AvoidPattern(\"TTGACA\"), # E.coli sig70 -35 site\n",
    "    AvoidPattern(\"TATAAT\"), # E.coli sig70 -10 site\n",
    "    AvoidPattern(\"TTGNNNNNNNNNNNNNNNNNNNNTATNNT\"), # E.coli sig70 promoter weak consensus,\n",
    "    AvoidPattern(\"TGGCACGNNNNTTGC\"), # E.coli sig54 promoter consensus\n",
    "    AvoidPattern(\"GAACTNNNNNNNNNNNNNNNNGTCNNA\"), # E.coli sig24 promoter consensus\n",
    "    AvoidPattern(\"AAAGA\"), # RBS\n",
    "    AvoidPattern(\"AGGAGG\"), # Shine-Dalgarno sequence or 2xArg bad codon\n",
    "    AvoidPattern(\"ATG\"), # Start codon\n",
    "    AvoidPattern(\"TTATNCACA\"), # DnaA binding sites\n",
    "    AvoidPattern(\"TGTGANNNNNNTCACANT\"), # CAP binding sites\n",
    "    AvoidPattern(\"NGCTNAGCN\"), # IS10 insertion site\n",
    "    AvoidPattern(\"GGGNNNNNCCC\"), # IS231 insertion site\n",
    "    AvoidPattern(\"(G{3,}[ATGC]{1,7}){3,}G{3,}\"), # G-quadruplex\n",
    "    AvoidPattern(\"GGGG\"), # G-quadruplex\n",
    "    AvoidChanges(location=(0, 6, 1)),\n",
    "    AvoidChanges(location=(len(linker_base_sequence) - 4, len(linker_base_sequence), 1))\n",
    "]\n",
    "problem = DnaOptimizationProblem(\n",
    "    sequence=linker_base_sequence,\n",
    "    constraints=constraints\n",
    ")\n",
    "problem.resolve_constraints()\n",
    "print(problem.constraints_text_summary())\n",
    "print(linker_base_sequence)\n",
    "print(problem.sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Group together parts and linkers from collections\n",
    "parts_linkers_collections = (\n",
    "    bsb.BASIC_BIOLEGIO_LINKERS[\"v0.1\"].values(),\n",
    "    bsb.BASIC_PROMOTER_PARTS[\"v0.2\"].values(),\n",
    "    bsb.BASIC_SEVA_PARTS[\"v0.1\"].values()\n",
    ")\n",
    "core_parts_linkers = []\n",
    "for part_linker_collection in parts_linkers_collections:\n",
    "    core_parts_linkers += list(part_linker_collection)\n",
    "# Add E.coli MG1655 genome sequence\n",
    "Entrez.email = \"hainesm6@gmail.com\"\n",
    "import os, ssl\n",
    "if (not os.environ.get('PYTHONHTTPSVERIFY', '') and\n",
    "getattr(ssl, '_create_unverified_context', None)):\n",
    "    ssl._create_default_https_context = ssl._create_unverified_context\n",
    "with Entrez.efetch(db=\"Nucleotide\", id=\"NZ_LR881938.1\", rettype=\"fasta\", retmode=\"text\") as handle:\n",
    "    mg1655 = SeqIO.read(handle, \"fasta\")\n",
    "    seqs = core_parts_linkers + [mg1655]\n",
    "path_to_seqs = Path.cwd().parents[0] / \"sequences\"\n",
    "bsb.export_sequences_to_file(\n",
    "    seqs,\n",
    "    Path.cwd().parents[0] / \"sequences\" / \"alternative_formats\" / \"fasta\" / \"basic_homology_sequences.fa\",\n",
    "    \"fasta\"\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "constraint:   0%|          | 0/22 [00:00<?, ?it/s, now=AvoidPattern[0-55](patter...]\n",
      "\n",
      "location:   0%|          | 0/1 [00:00<?, ?it/s, now=None]\u001b[A\u001b[A\n",
      "\n",
      "location:   0%|          | 0/1 [00:00<?, ?it/s, now=36-39(+)]\u001b[A\u001b[A\n",
      "\n",
      "                                                             \u001b[A\u001b[A\n",
      "\n",
      "location:   0%|          | 0/10 [00:00<?, ?it/s, now=36-39(+)]\u001b[A\u001b[A\n",
      "\n",
      "location:   0%|          | 0/10 [00:00<?, ?it/s, now=8-23]    \u001b[A\u001b[A\n",
      "\n",
      "                                                          \u001b[A\u001b[A\n",
      "\n",
      "location:   0%|          | 0/10 [00:00<?, ?it/s, now=8-23]\u001b[A\u001b[A\n",
      "\n",
      "location:  10%|█         | 1/10 [00:00<00:02,  3.91it/s, now=9-24]\u001b[A\u001b[A\n",
      "\n",
      "location:  10%|█         | 1/10 [00:00<00:02,  3.89it/s, now=9-24]\u001b[A\u001b[A\n",
      "\n",
      "location:  20%|██        | 2/10 [00:00<00:02,  3.89it/s, now=29-44]\u001b[A\u001b[A\n",
      "\n",
      "location:  20%|██        | 2/10 [00:00<00:02,  3.78it/s, now=29-44]\u001b[A\u001b[A\n",
      "\n",
      "location:  50%|█████     | 5/10 [00:00<00:01,  3.78it/s, now=38-53]\u001b[A\u001b[A\n",
      "\n",
      "location:  50%|█████     | 5/10 [00:00<00:01,  4.74it/s, now=38-53]\u001b[A\u001b[A\n",
      "\n",
      "location: 100%|██████████| 10/10 [00:01<00:00,  5.03it/s, now=38-53]\u001b[A\u001b[A\n",
      "\n",
      "                                                                                             ===> SUCCESS - all constraints evaluations pass\n",
      "✔PASS ┍ EnforceMeltingTemperature[13-34]\n",
      "      │ Tm = 51.6\n",
      "✔PASS ┍ AvoidPattern[0-55](pattern:EcoRI(GAATTC))\n",
      "      │ Passed. Pattern not found !\n",
      "✔PASS ┍ AvoidPattern[0-55](pattern:SpeI(ACTAGT))\n",
      "      │ Passed. Pattern not found !\n",
      "✔PASS ┍ AvoidPattern[0-55](pattern:XbaI(TCTAGA))\n",
      "      │ Passed. Pattern not found !\n",
      "✔PASS ┍ AvoidPattern[0-55](pattern:PstI(CTGCAG))\n",
      "      │ Passed. Pattern not found !\n",
      "✔PASS ┍ AvoidPattern[0-55](pattern:BsaI(GGTCTC))\n",
      "      │ Passed. Pattern not found !\n",
      "✔PASS ┍ AvoidPattern[0-55](pattern:BsmBI(CGTCTC))\n",
      "      │ Passed. Pattern not found !\n",
      "✔PASS ┍ AvoidPattern[0-55](pattern:TTGACA)\n",
      "      │ Passed. Pattern not found !\n",
      "✔PASS ┍ AvoidPattern[0-55](pattern:TATAAT)\n",
      "      │ Passed. Pattern not found !\n",
      "✔PASS ┍ AvoidPattern[0-55](pattern:TTGNNNNNNNNNNNNNNNNNNNNTATNNT)\n",
      "      │ Passed. Pattern not found !\n",
      "✔PASS ┍ AvoidPattern[0-55](pattern:TGGCACGNNNNTTGC)\n",
      "      │ Passed. Pattern not found !\n",
      "✔PASS ┍ AvoidPattern[0-55](pattern:GAACTNNNNNNNNNNNNNNNNGTCNNA)\n",
      "      │ Passed. Pattern not found !\n",
      "✔PASS ┍ AvoidPattern[0-55](pattern:AAAGA)\n",
      "      │ Passed. Pattern not found !\n",
      "✔PASS ┍ AvoidPattern[0-55](pattern:AGGAGG)\n",
      "      │ Passed. Pattern not found !\n",
      "✔PASS ┍ AvoidPattern[0-55](pattern:ATG)\n",
      "      │ Passed. Pattern not found !\n",
      "✔PASS ┍ AvoidPattern[0-55](pattern:TTATNCACA)\n",
      "      │ Passed. Pattern not found !\n",
      "✔PASS ┍ AvoidPattern[0-55](pattern:TGTGANNNNNNTCACANT)\n",
      "      │ Passed. Pattern not found !\n",
      "✔PASS ┍ AvoidPattern[0-55](pattern:NGCTNAGCN)\n",
      "      │ Passed. Pattern not found !\n",
      "✔PASS ┍ AvoidPattern[0-55](pattern:GGGNNNNNCCC)\n",
      "      │ Passed. Pattern not found !\n",
      "✔PASS ┍ AvoidPattern[0-55](pattern:(G{3,}[ATGC]{1,7}){3,}G{3,})\n",
      "      │ Passed. Pattern not found !\n",
      "✔PASS ┍ AvoidPattern[0-55](pattern:GGGG)\n",
      "      │ Passed. Pattern not found !\n",
      "✔PASS ┍ AvoidChanges[0-6(+)]\n",
      "      │ Enforced by nucleotides restrictions\n",
      "✔PASS ┍ AvoidChanges[51-55(+)]\n",
      "      │ Enforced by nucleotides restrictions\n",
      "✔PASS ┍ AvoidMatches[0-55]\n",
      "      │ Failed - 0 matches at []\n",
      "\n",
      "\n",
      "GGCTCGCTCCACCTGCGTCTATCGTACAGTGCTAAAATGGCAGCAGATAGTGTCC\n",
      "GGCTCGCTCCACTCGCGTCTATCGTACAGTGATAAAATCGGAGGAGAGAGTGTCC\n"
     ]
    }
   ],
   "source": [
    "constraints += [AvoidMatches(15, bowtie_index=path_to_seqs / \"alternative_formats\" / \"bowtie_indexes\" / \"2021-07-22_basic_homology\" / \"basic_homology\", mismatches=1)]\n",
    "problem = DnaOptimizationProblem(\n",
    "    sequence=linker_base_sequence,\n",
    "    constraints=constraints\n",
    ")\n",
    "problem.resolve_constraints()\n",
    "assert problem.sequence[:len(\"GGCTCG\")] == \"GGCTCG\"\n",
    "assert problem.sequence[-1*len(\"GTCC\"):] == \"GTCC\"\n",
    "print(problem.constraints_text_summary())\n",
    "print(linker_base_sequence)\n",
    "print(problem.sequence)"
   ]
  },
  {
   "source": [
    "## Aims and objectives for the cell/s below\n",
    "\n",
    "- Evaluate the generated linker sequence:\n",
    "     - [ ] What is the sequence of the suffix and prefix long linker oligonucleotides?\n",
    "     - [ ] What is the sequence of the suffix and prefix adapter linker oligonucleotides?\n",
    "     - [ ] What is the melting temperature of the overlap region?"
   ],
   "cell_type": "markdown",
   "metadata": {}
  }
 ]
}